File: src/main/java/com/paragon/client/ui/configuration/old/impl/setting/KeybindComponent.java
package com.paragon.client.ui.configuration.old.impl.setting;

import com.paragon.Paragon;
import com.paragon.api.event.client.SettingUpdateEvent;
import com.paragon.api.setting.Bind;
import com.paragon.api.setting.Setting;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.api.util.render.font.FontUtil;
import com.paragon.client.ui.configuration.old.impl.module.ModuleButton;
import net.minecraft.util.text.TextFormatting;
import org.lwjgl.input.Keyboard;
import org.lwjgl.opengl.GL11;

import java.awt.*;

public class KeybindComponent extends SettingComponent<Bind> {

    private boolean isListening = false;

    public KeybindComponent(ModuleButton moduleButton, Setting<Bind> setting, float offset, float height) {
        super(moduleButton, setting, offset, height);
    }

    @Override
    public void renderSetting(int mouseX, int mouseY) {
        RenderUtil.drawRect(getModuleButton().getPanel().getX(), getModuleButton().getOffset() + getOffset(), getModuleButton().getPanel().getWidth(), getHeight(), isMouseOver(mouseX, mouseY) ? new Color(23, 23, 23).brighter().getRGB() : new Color(23, 23, 23).getRGB());

        GL11.glPushMatrix();
        GL11.glScalef(0.65f, 0.65f, 0.65f);

        {
            float scaleFactor = 1 / 0.65f;
            FontUtil.drawStringWithShadow(getSetting().getName(), (getModuleButton().getPanel().getX() + 5) * scaleFactor, (getModuleButton().getOffset() + getOffset() + 4.5f) * scaleFactor, -1);

            float side = (getModuleButton().getPanel().getX() + getModuleButton().getPanel().getWidth() - (FontUtil.getStringWidth(isListening ? " ..." : " " + getSetting().getValue().getButtonName()) * 0.65f) - 5) * scaleFactor;
            FontUtil.drawStringWithShadow(TextFormatting.GRAY + (isListening ? " ..." : " " + getSetting().getValue().getButtonName()), side, (getModuleButton().getOffset() + getOffset() + 4.5f) * scaleFactor, -1);
        }

        GL11.glPopMatrix();

        super.renderSetting(mouseX, mouseY);
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
        if (isListening) {
            isListening = false;

            getSetting().getValue().setDevice(Bind.Device.MOUSE);
            getSetting().getValue().setButtonCode(mouseButton);

            return;
        }

        if (isMouseOver(mouseX, mouseY) && mouseButton == 0) {
            // Set listening
            isListening = !isListening;
            SettingUpdateEvent settingUpdateEvent = new SettingUpdateEvent(getSetting());
            Paragon.INSTANCE.getEventBus().post(settingUpdateEvent);
        }

        if (isExpanded()) {
            getSettingComponents().forEach(settingComponent -> {
                if (settingComponent.getSetting().isVisible()) {
                    settingComponent.mouseClicked(mouseX, mouseY, mouseButton);
                }
            });
        }

        super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    public void keyTyped(char typedChar, int keyCode) {
        if (isListening) {
            isListening = false;

            getSetting().getValue().setDevice(Bind.Device.KEYBOARD);

            if (keyCode == Keyboard.KEY_DELETE || keyCode == Keyboard.KEY_BACK) {
                getSetting().getValue().setButtonCode(0);
                return;
            }

            getSetting().getValue().setButtonCode(keyCode);
        }

        if (isExpanded()) {
            getSettingComponents().forEach(settingComponent -> settingComponent.keyTyped(typedChar, keyCode));
        }
    }

    @Override
    public float getAbsoluteHeight() {
        float subsettingHeight = 0;

        for (SettingComponent<?> settingComponent : getSettingComponents()) {
            subsettingHeight += settingComponent.getHeight();
        }

        return isExpanded() ? getHeight() + subsettingHeight : getHeight();
    }
}

File: src/main/kotlin/com/paragon/client/ui/configuration/old/impl/setting/KeybindComponent.kt
package com.paragon.client.ui.configuration.old.impl.setting

import com.paragon.Paragon
import com.paragon.api.event.client.SettingUpdateEvent
import com.paragon.api.setting.Bind
import com.paragon.api.setting.Setting
import com.paragon.api.util.render.RenderUtil.drawRect
import com.paragon.api.util.render.font.FontUtil.drawStringWithShadow
import com.paragon.api.util.render.font.FontUtil.getStringWidth
import com.paragon.client.ui.configuration.old.impl.module.ModuleButton
import net.minecraft.util.text.TextFormatting
import org.lwjgl.input.Keyboard
import org.lwjgl.opengl.GL11
import java.awt.Color
import java.util.function.Consumer

class KeybindComponent(moduleButton: ModuleButton, setting: Setting<Bind?>, offset: Float, height: Float) : SettingComponent<Bind?>(moduleButton, setting, offset, height) {
    private var isListening = false
    override fun renderSetting(mouseX: Int, mouseY: Int) {
        drawRect(moduleButton.panel.x, moduleButton.offset + offset, moduleButton.panel.width, height, if (isMouseOver(mouseX, mouseY)) Color(23, 23, 23).brighter().rgb else Color(23, 23, 23).rgb)
        GL11.glPushMatrix()
        GL11.glScalef(0.65f, 0.65f, 0.65f)
        run {
            val scaleFactor = 1 / 0.65f
            drawStringWithShadow(setting.name, (moduleButton.panel.x + 5) * scaleFactor, (moduleButton.offset + offset + 4.5f) * scaleFactor, -1)
            val side = (moduleButton.panel.x + moduleButton.panel.width - getStringWidth(if (isListening) " ..." else " " + setting.value!!.getButtonName()) * 0.65f - 5) * scaleFactor
            drawStringWithShadow(TextFormatting.GRAY.toString() + if (isListening) " ..." else " " + setting.value!!.getButtonName(), side, (moduleButton.offset + offset + 4.5f) * scaleFactor, -1)
        }
        GL11.glPopMatrix()
        super.renderSetting(mouseX, mouseY)
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        if (isListening) {
            isListening = false
            setting.value!!.device = Bind.Device.MOUSE
            setting.value!!.buttonCode = mouseButton
            return
        }
        if (isMouseOver(mouseX, mouseY) && mouseButton == 0) {
            // Set listening
            isListening = !isListening
            val settingUpdateEvent = SettingUpdateEvent(setting)
            Paragon.INSTANCE.eventBus.post(settingUpdateEvent)
        }
        if (isExpanded) {
            settingComponents.forEach(Consumer { settingComponent: SettingComponent<*> ->
                if (settingComponent.setting.isVisible()) {
                    settingComponent.mouseClicked(mouseX, mouseY, mouseButton)
                }
            })
        }
        super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    override fun keyTyped(typedChar: Char, keyCode: Int) {
        if (isListening) {
            isListening = false
            setting.value!!.device = Bind.Device.KEYBOARD
            if (keyCode == Keyboard.KEY_DELETE || keyCode == Keyboard.KEY_BACK) {
                setting.value!!.buttonCode = 0
                return
            }
            setting.value!!.buttonCode = keyCode
        }
        if (isExpanded) {
            settingComponents.forEach(Consumer { settingComponent: SettingComponent<*> -> settingComponent.keyTyped(typedChar, keyCode) })
        }
    }
}